pkg/operator/csr/csr_approver.go

```diff
@@ -0,0 +1,272 @@
+	CSRDenied    CSRApprovalDecision = "Denied"
+	CSRNoOpinion CSRApprovalDecision = "NoOpinion"
+)
+
+type CSRApprover interface {
+	Approve(csrObj *certapiv1.CertificateSigningRequest, x509CSR *x509.CertificateRequest) (approvalStatus CSRApprovalDecision, denyReason string, err error)
+}
+
+type csrApproverController struct {
+	csrClient certv1client.CertificateSigningRequestInterface
+	csrLister certv1listers.CertificateSigningRequestLister
+
+	csrFilter   CSRFilter
+	csrApprover CSRApprover
+}
+
+// NewCSRApproverController returns a controller that is observing the CSR API
+// for a CSR of a given name. If such a CSR exists, it runs the `csrApprover.Approve()`
+// against it and either denies, approves or leaves the CSR.
+func NewCSRApproverController(
+	controllerName string,
+	operatorClient v1helpers.OperatorClient,
+	csrClient certv1client.CertificateSigningRequestInterface,
+	csrInformers certv1informers.CertificateSigningRequestInformer,
+	csrFilter CSRFilter,
+	csrApprover CSRApprover,
+	eventsRecorder events.Recorder,
+) factory.Controller {
+	c := &csrApproverController{
+		csrClient:   csrClient,
+		csrLister:   csrInformers.Lister(),
+		csrFilter:   csrFilter,
+		csrApprover: csrApprover,
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(csrInformers.Informer()).
+		ToController(
+			"WebhookAuthenticatorCertApprover_"+controllerName,
+			eventsRecorder.WithComponentSuffix("webhook-authenticator-cert-approver-"+controllerName),
+		)
+}
+
+func (c *csrApproverController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	csrList, err := c.csrLister.List(labels.Everything())
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	for _, csr := range csrList {
+		if !c.csrFilter.Matches(csr) {
+			continue
+		}
+
+		if approved, denied := getCertApprovalCondition(&csr.Status); approved || denied {
+			return nil
+		}
+
+		csrCopy := csr.DeepCopy()
+
+		csrPEM, _ := pem.Decode(csr.Spec.Request)
+		if csrPEM == nil {
+			return fmt.Errorf("failed to PEM-parse the CSR block in .spec.request: no CSRs were found")
+		}
+
+		x509CSR, err := x509.ParseCertificateRequest(csrPEM.Bytes)
+		if err != nil {
+			return fmt.Errorf("failed to parse the CSR bytes: %v", err)
+		}
+
+		if x509CSR.Subject.CommonName == csr.Spec.Username {
+			c.denyCSR(ctx, csrCopy, "IllegitimateRequester", "requester cannot request certificates for themselves", syncCtx.Recorder())
+		}
+
+		csrDecision, denyReason, err := c.csrApprover.Approve(csr, x509CSR)
+		if err != nil {
+			return c.denyCSR(ctx, csrCopy, "CSRApprovingFailed", fmt.Sprintf("there was an error during CSR approval: %v", err), syncCtx.Recorder())
+		}
+
+		switch csrDecision {
+		case CSRDenied:
+			return c.denyCSR(ctx, csrCopy, "CSRDenied", denyReason, syncCtx.Recorder())
+		case CSRApproved:
+			csrCopy.Status.Conditions = append(csrCopy.Status.Conditions,
```

Comment:
Perhaps it makes sense to extract this into a function as well (as you did with `CSRDenied` case)?

Reply:
it's single use so I did not want to do that, but there's no hurt in doing so :+1: